py3status/modules/google_calendar.py

```diff
@@ -0,0 +1,521 @@
+                event_str = dict((k, v.lower()) if v is not None else (k, v)
+                                 for k, v in event_str.items())
+
+            event_formatted = self.py3.safe_format(
+                curr_event_format,
+                {'summary': event_str['summary'],
+                 'location': event_str['location'],
+                 'description': event_str['description'],
+                 'start_time': event_str['start_time'],
+                 'end_time': event_str['end_time'],
+                 'start_date': event_str['start_date'],
+                 'end_date': event_str['end_date'],
+                 'time_to': time_to_composite})
+
+            self.py3.composite_update(event_formatted, {'index': index})
+            responses.append(event_formatted)
+
+            index += 1
+
+        format_separator = self.py3.safe_format(self.format_separator)
+        self.py3.composite_update(format_separator, {'index': 'sep'})
+        responses = self.py3.composite_join(format_separator, responses)
+
+        return {'events': responses}
+
+    def google_calendar(self):
+        """
+        The method that outputs the response.
+
+        First, we check credential authorization. If no authorization, we display an error message,
+        and try authorizing again in 5 seconds.
+
+        Otherwise, we fetch the events, build the response, and output the resulting composite.
+        """
+        if not self.is_authorized:
+            self.is_authorized = self._authorize_credentials()
+
+            return {'cached_until': self.py3.time_in(self.cache_timeout),
+                    'full_text': self.py3.safe_format(self.format)}
+        else:
+            if not self.no_update:
+                self.events = self._get_events()
+
+            return {
+                'cached_until': self.py3.time_in(self.cache_timeout),
+                'composite': self.py3.safe_format(self.format, self._build_response())
+            }
+
+    def on_click(self, event):
+        if self.is_authorized and self.events is not None:
+            """
+            If button_refresh is clicked, we allow the events to be updated if the
+            last event update occured at least 1 second ago. This prevents a bug that
+            can crash py3status since refreshing the module too fast results in
+            incomplete event information being fetched as _get_events() is called
+            repeatedly.
+
+            Otherwise, we disable event updates.
+            """
+            try:
+                self.no_update = True
+                self.button = event['button']
+                button_index = event['index']
+
+                if self.button == self.button_refresh:
+                    now = datetime.datetime.now()
+                    diff = (now - self.last_update).seconds
+                    if diff > 1:
+                        self.no_update = False
+                elif button_index != 'sep':
```

Comment:
Maybe we can do this to get rid of the `sep` indexes and to clean up a bit.
``` 
        try:
            button_index = event['index']
        except IndexError:
            self.py3.prevent_refresh()
```

Reply:
Before adding `sep`, clicking on a separator sometimes returns an erroneous index. For example, if I click on the first separator, `event['index']` is 2.  If I click on the third separator, `event['index']` is 8 (which is outside the range of events). Why, i'm not sure. Here's some sample output with `sep` removed:

```
[  
   {  
      'full_text':'testasd (test)',
      'index':0,
      'separator':False,
      'separator_block_width':0,
      'cached_until':1509090350.0
   },
   {  
      'full_text':'|',
      'separator':False,
      'separator_block_width':0
   },
   {  
      'full_text':'CSI 3105 A Lecture (Morisset 205)',
      'index':1,
      'separator':False,
      'separator_block_width':0
   },
...
```

You can see there's no index on the separator, but `event['index'] will contain some erroneous index if the separator is clicked on.

Reply:
I think that was before you had the separators go through the format... eg
```
responses.append({'full_text':' {} '.format(self.format_separator), 'index': 'sep'})
```
and that you had the former separators at end of the loop too... so they gets an `index` too... is why you may have `2` and even `8` too for `index`. We fixed that with `composite_join`ing the formatted events with the formatted separators.

Reply:
Sorry, I should have mentioned that this is based on trying your suggestion remove `sep` and `except`. Even with the current approach to separators, I get those weird indexes.

Reply:
Cool. I'm heading out right now (3:15am here). I'll check it out when I get back.

Reply:
Home. What happen when you remove this too?
```
self.py3.composite_update(format_separator, {'index': 'sep'})
```

Reply:
Same thing. That output I posted above + the strange index behavior is with that line removed.

Reply:
Can you add `self.py3.logs` and push so I can see what you're looking at... I don't see it here... or I'm logging things in wrong place.

Reply:
Yup, done. Just click on some separators and check the log.

Reply:
I guess it's an expected behavior... [#compostes](http://py3status.readthedocs.io/en/latest/writing_modules.html#composites) 👍

Reply:
note to self: RTFM

Reply:
Maybe we could check `sep` early so we don't worry about checking them in the `if-elif-else.`
```
                self.button = event['button']
                button_index = event['index']
                if button_index == 'sep':
                    self.py3.prevent_refresh()
```
This could use `self.py3.prevent_refresh()` too as nothing should change since we're just opening an event link.
```
    self.py3.command_run('xdg-open ' + self.event_urls[button_index])
    self.py3.prevent_refresh()
```